pkg/csi/service/vanilla/controller.go

```diff
@@ -1243,14 +1244,75 @@ func (c *controller) CreateSnapshot(ctx context.Context, req *csi.CreateSnapshot
 				"cannot snapshot migrated vSphere volume. :%q", volumeID)
 		}
 
-		// Query Capacity in MB for block volume snapshot
-		snapshotSizeInMB, err := utils.QueryCapacityForBlockVolumeSnapshotUtil(
+		// Query capacity in MB and datastore url for block volume snapshot
+		snapshotSizeInMB, datastoreUrl, err := utils.QueryCapacityAndDsUrlForVolumeUtil(
 			ctx, c.manager.VolumeManager, volumeID, common.BlockVolumeType)
 		if err != nil {
 			return nil, err
 		}
 		volumeType = prometheus.PrometheusBlockVolumeType
 
+		// Check if snapshots number of this volume reaches the granular limit on VSAN/VVOL
+		maxSnapshotsPerBlockVolume := c.manager.CnsConfig.Snapshot.GlobalMaxSnapshotsPerBlockVolume
+		log.Infof("The limit of the maximum number of snapshots per block volume is "+
+			"set to the global maximum (%v) by default.", maxSnapshotsPerBlockVolume)
+		if c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVSAN != 0 ||
+			c.manager.CnsConfig.Snapshot.GranularMaxSnapshotsPerBlockVolumeInVVOL != 0 {
+			// query the datastore type of the source volume
+			vCenter, err := cnsvsphere.GetVirtualCenterManager(ctx).GetVirtualCenter(ctx, c.manager.VcenterConfig.Host)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to get vCenter when checking if the number of snapshots on the source volume "+
+						"reaches the configured maximum. err: %v", err)
+			}
+
+			dsInfo, err := cnsvsphere.GetDatastoreInfoByURL(
+				ctx, vCenter, c.manager.CnsConfig.Global.ClusterID, datastoreUrl)
+			if err != nil {
+				return nil, logger.LogNewErrorCodef(log, codes.Internal,
+					"failed to get the datastore info corresponding to URL %q when checking if the number of "+
+						"snapshots on the source volume reaches the configured maximum. err: %v", datastoreUrl, err)
+			}
+
+			dsType, err := dsInfo.Type(ctx)
```

Comment:
The dsURL will have "vvol" or "vsan" based on datastore type, you can use this to avoid making another call to vc.